iib/workers/tasks/build.py

```diff
@@ -337,6 +340,150 @@ def _get_resolved_image(pull_spec):
+    run_cmd(
+        ['podman', 'create', '--cidfile', cidfile, from_index],
+        exc_msg='Failed to create index image container',
+    )
+    with open(cidfile, 'r') as f:
+        cid = f.read()
+    os.remove(cidfile)
+
+    try:
+        data = json.loads(
+            run_cmd(
+                ['podman', 'image', 'inspect', from_index], exc_msg='Failed to fetch image metadata'
+            )
+        )
+        db_path = data[0]['Config']['Labels']['operators.operatorframework.io.index.database.v1']
+        run_cmd(
+            ['podman', 'cp', f'{cid}:{db_path}', base_dir],
+            exc_msg='Failed to copy DB file from the container',
+        )
+    finally:
+        run_cmd(['podman', 'rm', cid], exc_msg='Failed to delete the container')
+    local_path = os.path.join(base_dir, db_path.split('/')[-1])
+
+    return local_path
+
+
+def _serve_index_registry(db_path, max_port_tries=100, start_port=50051, init_wait_time=1):
+    """
+    Locally start OPM registry service, which can be communicated with using gRPC queries.
+
+    Due to IIB's paralellism, the service can run multiple times, which could lead to port
+    binding conflicts. Resolution of port conflicts is handled in this function as well.
+
+    :param str db_path: path to index database containing the registry data
+    :param int max_port_tries: Maximum number of ports to try before giving up
+    :param int start_port: First port number ot try. Subsequent ports are incrementations
+    :param int init_wait_time: Time to wait between starting the service and checking its output
+    :return: tuple containing port number of the running service and the runnung Popen object
+    :rtype: (int, Popen)
+    :raises IIBError: if all tried ports are in use, or the command failed for another reason
+    """
+    port = start_port
+    for i in range(max_port_tries):
+        cmd = ['opm', 'registry', 'serve', '-p', str(port), '-d', db_path]
+        rpc_proc = subprocess.Popen(
+            cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True
+        )
+        time.sleep(init_wait_time)
+        ret = rpc_proc.poll()
+        if not ret:
+            log.debug('Running the command "%s"', ' '.join(cmd))
+            break
+        else:
+            stderr = rpc_proc.stderr.read()
+            if 'address already in use' in stderr:
+                if i == max_port_tries - 1:
+                    err_msg = f'No free port has been found after {max_port_tries} attempts.'
+                    log.error(err_msg)
+                    raise IIBError(err_msg)
+                log.info('Port %s is in use, trying another.', port)
+                port += 1
+            else:
+                raise IIBError('Command "%s" has failed with error "%s"', ' '.join(cmd), stderr)
+
+    return (port, rpc_proc)
+
+
+@retry(attempts=2, wait_on=IIBError, logger=log)
+def _get_present_bundles(from_index, base_dir, grpc_max_tries=5, grpc_wait_time=1):
```

Comment:
How about using `get_worker_config().iib_total_attempts` for grpc_max_tries instead of static values?

Reply:
It might be useful to make the wait_time configurable too?

Reply:
I have added these values to config